pkg/controller/master/node/cpu_manager_controller.go

```diff
@@ -0,0 +1,322 @@
+)
+
+type CPUManagerPolicy string
+type CPUManagerStatus string
+
+type CPUManagerUpdateStatus struct {
+	Policy  CPUManagerPolicy `json:"policy"`
+	Status  CPUManagerStatus `json:"status"`
+	JobName string           `json:"jobName,omitempty"`
+}
+
+// cpuManagerNodeHandler updates cpu manager status of a node in its annotations, so that
+// we can tell whether the node is under modifing cpu manager policy or not and its current policy.
+type cpuManagerNodeHandler struct {
+	appCache   catalogv1.AppCache
+	nodeCache  ctlcorev1.NodeCache
+	nodeClient ctlcorev1.NodeClient
+	jobClient  ctlbatchv1.JobClient
+	namespace  string
+}
+
+// CPUManagerRegister registers the node controller
+func CPUManagerRegister(ctx context.Context, management *config.Management, options config.Options) error {
+	app := management.CatalogFactory.Catalog().V1().App()
+	job := management.BatchFactory.Batch().V1().Job()
+	node := management.CoreFactory.Core().V1().Node()
+
+	cpuManagerNodeHandler := &cpuManagerNodeHandler{
+		appCache:   app.Cache(),
+		jobClient:  job,
+		nodeCache:  node.Cache(),
+		nodeClient: node,
+		namespace:  options.Namespace,
+	}
+
+	node.OnChange(ctx, CPUManagerControllerName, cpuManagerNodeHandler.OnNodeChanged)
+	job.OnChange(ctx, CPUManagerControllerName, cpuManagerNodeHandler.OnJobChanged)
+
+	return nil
+}
+
+func (h *cpuManagerNodeHandler) OnNodeChanged(_ string, node *corev1.Node) (*corev1.Node, error) {
+	if node == nil || node.Annotations == nil || node.DeletionTimestamp != nil {
+		return node, nil
+	}
+
+	annot, found := node.Annotations[util.AnnotationCPUManagerUpdateStatus]
+	if !found {
+		return node, nil
+	}
+
+	// do nothing if the node is in witness role
+	if _, found := node.Labels[HarvesterWitnessNodeLabelKey]; found {
+		return node, nil
+	}
+
+	cpuManagerStatus, err := GetCPUManagerUpdateStatus(annot)
+	if err != nil {
+		logrus.WithField("node_name", node.Name).WithError(err).Error("Skip update cpu manager policy, failed to retrieve cpu-manager-update-status from annotation")
+		return node, nil
+	}
+	// do nothing if status is not requested
+	if cpuManagerStatus.Status != CPUManagerRequestedStatus {
+		return node, nil
+	}
+
+	newNode := node.DeepCopy()
+	job, err := h.submitJob(cpuManagerStatus, node)
+	logrus.WithFields(logrus.Fields{
+		"node_name": node.Name,
+		"job_name":  job.Name,
+		"policy":    cpuManagerStatus.Policy,
+	}).Info("Submit cpu manager job")
+	if err != nil {
+		logrus.WithField("node_name", node.Name).WithError(err).Error("Submit cpu manager job failed")
+		setUpdateStatus(newNode, toFailedStatus(cpuManagerStatus))
+	} else {
+		setUpdateStatus(newNode, toRunningStatus(cpuManagerStatus, job))
+	}
+
+	if reflect.DeepEqual(newNode, node) {
+		return node, nil
+	}
+	return h.nodeClient.Update(newNode)
+}
+
+func (h *cpuManagerNodeHandler) OnJobChanged(_ string, job *batchv1.Job) (*batchv1.Job, error) {
+	if job == nil || job.Labels == nil || job.DeletionTimestamp != nil {
```

Comment:
When a job is running but then deleted by user manually, are there any chance the node is updated with successful/failed?